squad/ci/models.py

```diff
@@ -68,60 +68,58 @@ def fetch(self, test_job):
         results = implementation.fetch(test_job)
+        if not results:
+            test_job.save()
+            return
 
-        if results:
-            # create TestRun
-            status, completed, metadata, tests, metrics, logs = results
+        status, completed, metadata, tests, metrics, logs = results
 
-            test_job.job_status = status
-            if not completed:
-                tests = {}
-                metrics = {}
-                test_job.can_resubmit = True
+        test_job.job_status = status
+        if not completed:
+            tests = {}
+            metrics = {}
+            test_job.can_resubmit = True
 
-            if completed and not tests and not metrics:
-                # test job produced no results
-                # mark it incomplete
-                completed = False
-                test_job.can_resubmit = True
+        if completed and not tests and not metrics:
+            # test job produced no results
+            # mark it incomplete
+            completed = False
+            test_job.can_resubmit = True
 
-            metadata['job_id'] = test_job.job_id
-            metadata['job_status'] = test_job.job_status
-            if test_job.url is not None:
-                metadata['job_url'] = test_job.url
+        metadata['job_id'] = test_job.job_id
+        metadata['job_status'] = test_job.job_status
+        if test_job.url is not None:
+            metadata['job_url'] = test_job.url
 
-            try:
-                receive = ReceiveTestRun(test_job.target, update_project_status=False)
-                testrun = receive(
-                    version=test_job.target_build.version,
-                    environment_slug=test_job.environment,
-                    metadata_file=json.dumps(metadata),
-                    tests_file=json.dumps(tests),
-                    metrics_file=json.dumps(metrics),
-                    log_file=logs,
-                    completed=completed,
-                )
-                test_job.testrun = testrun
-                test_job.fetched = True
-                test_job.fetched_at = timezone.now()
-                test_job.save()
-
-                self.__postprocess_testjob__(test_job)
+        try:
+            # create TestRun
+            receive = ReceiveTestRun(test_job.target, update_project_status=False)
+            testrun = receive(
+                version=test_job.target_build.version,
+                environment_slug=test_job.environment,
+                metadata_file=json.dumps(metadata),
+                tests_file=json.dumps(tests),
+                metrics_file=json.dumps(metrics),
+                log_file=logs,
+                completed=completed,
+            )
+            test_job.testrun = testrun
+        except InvalidMetadata as exception:
+            test_job.failure = str(exception)
+
+        # mark test job as fetched to prevent resubmission
+        # on next fetch attempt
+        test_job.fetched = True
```

Comment:
I think this block should be inside 'except' block. Otherwise you will mark job as fetched even if creating TestRun fails. I'm not sure this is a good idea.